File: WZXKaiYanDemo/ViewController.m
//
//  ViewController.m
//  WZXKaiYanDemo
//
//  Created by wordoor－z on 16/1/25.
//  Copyright © 2016年 wzx. All rights reserved.
//

#import "ViewController.h"
#import "MyCell.h"

@interface ViewController ()<UITableViewDataSource,UITableViewDelegate>

/**
 *  数据数组
 */
@property(nonatomic,strong) NSMutableArray * dataArray;

/**
 *  tableview
 */
@property(nonatomic,strong) UITableView * tableView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self preData];
    [self createTableView];
}

//准备数据
- (void)preData
{
    _dataArray = [[NSMutableArray alloc]init];
    for (int i = 0; i < 20; i++)
    {
        UIImage * image = [UIImage imageNamed:[NSString stringWithFormat:@"%d",i]];
        [_dataArray addObject:image];
    }
}

//创建tableView
- (void)createTableView
{
    _tableView = [[UITableView alloc]initWithFrame:CGRectMake(0, 64, self.view.frame.size.width, self.view.frame.size.height - 64) style:UITableViewStylePlain];
    _tableView.delegate = self;
    _tableView.dataSource = self;
    _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    [self.view addSubview:_tableView];
}

#pragma mark ---------tableView-----------
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return _dataArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    MyCell * cell = [tableView dequeueReusableCellWithIdentifier:@"myCell"];
    if (cell == nil)
    {
        cell = [[MyCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"myCell"];
    }
    return cell;
}

//在willDisplayCell里面处理数据能优化tableview的滑动流畅性
- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    MyCell * myCell = (MyCell *)cell;
    
    [myCell setImg:_dataArray[indexPath.row]];
    
    [myCell cellOffset];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return self.view.frame.size.height/3.0;
}


- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    // visibleCells 获取界面上能显示出来了cell
    NSArray<MyCell *> *array = [self.tableView visibleCells];
    
    //enumerateObjectsUsingBlock 类似于for，但是比for更快
    [array enumerateObjectsUsingBlock:^(MyCell * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        
        [obj cellOffset];
        
    }];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: KaiYan/ViewController.swift
//
//  ViewController.swift
//  KaiYan
//
//  Created by wzxjiang on 2017/5/15.
//  Copyright © 2017年 wzxjiang. All rights reserved.
//

import UIKit

class Cell: UITableViewCell {
    var backgroundImageView: UIImageView!
    
    var titleLabel: UILabel!

    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        createUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func createUI() {
        self.clipsToBounds = true
        
        backgroundImageView =
            UIImageView(frame: CGRect(x: 0, y: -0.5 * Cell.rowHeight, width: UIScreen.main.bounds.width, height: 2 * Cell.rowHeight))
        backgroundImageView.contentMode = .scaleAspectFill
        
        titleLabel = UILabel()
        titleLabel.textColor = .white
        titleLabel.font = UIFont.boldSystemFont(ofSize: 25)
        titleLabel.text = "[标题]"
        
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        
        contentView.addSubview(backgroundImageView)
        contentView.addSubview(titleLabel)
        
        contentView.addConstraints([
             NSLayoutConstraint(item: titleLabel, attribute: .centerX, relatedBy: .equal, toItem: contentView, attribute: .centerX, multiplier: 1, constant: 0),
             NSLayoutConstraint(item: titleLabel, attribute: .centerY, relatedBy: .equal, toItem: contentView, attribute: .centerY, multiplier: 1, constant: 0)
            ])
    }
    
    func move() {
        DispatchQueue.global().async {
            
            guard let superView = self.superview else { return }
            
            let offsetY = self.convert(self.bounds, to: self.window).midY - superView.center.y
            
            let offset = (offsetY / superView.frame.size.height) * Cell.rowHeight
            
            let trans = CGAffineTransform(translationX: 0, y: -offset)
            
            DispatchQueue.main.async {
                self.backgroundImageView.transform = trans
            }
        }
    }
    
    class var rowHeight: CGFloat { return 250.0 }
}

class ViewController: UITableViewController {

    let datas = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
    
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.rowHeight = Cell.rowHeight
        tableView.register(Cell.self, forCellReuseIdentifier: "Cell")
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        guard let cell = cell as? Cell else {
            return
        }
        
        cell.backgroundImageView.image = UIImage(named: "\(datas[indexPath.row]).png")
        
        cell.move()
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "Cell") else {
            return Cell()
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return datas.count
    }
    
    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
        tableView.visibleCells.map {
            return $0 as? Cell
        }.forEach {
            $0?.move()
        }
    }
}

